### PR TITLE
Shallow routing

### LINKED ISSUE #2673: Support for Shallow Routing (from Next.js)
### Describe the problem

See: https://nextjs.org/docs/routing/shallow-routing

> Shallow routing allows you to change the URL without running data fetching methods again, (...) [it] only works for same page URL changes.

### Describe the proposed solution

```js
import { goto } from '$app/navigation';
await goto(href, { shallow: true });
```

plus

```svelte
<a sveltekit:shallow {href}></a>
```

These would update the `page` store, but `load` would not run. (Only if `href` does not link to a new page)

### Alternatives considered

Using the History API directly:
```js
history.replaceState(history.state, document.title, href)
```

But doing this does not update the `page` store (and it's ugly).

### Importance

would make my life easier

### Additional Information

I have this `OrgTree` component which lazy loads it's items as you expand the groups:

When you click an item the URI is updated using shallow routing, leaving all the loaded tree data intact.

Using Svelte Kit, `load` fetches the initial tree data, but, after navigating, the initial data is loaded again, reseting the whole tree.

Related issues: #1930, #969

### PR DESCRIPTION
**Pull request by Rich-Harris**
Closes #2673. This is a continuation of #9847, on top of the `version-2` branch. Copying the TODOs from that PR:

- [x] Add `App.PageState` and other types
- [x] remove `goto(url, { state })`
- [x] intercept `history.pushState` and `history.replaceState` to discourage their use
- [x] `invalidateAll()` should probably use `$page.url.href` rather than `location.href`, since those two things can now be different
- [x] ~~fix bugginess around reloads. If you push state and refresh the page, you go to pushed URL (as intended), and hitting the back button will take you to the page you came from, but only because there's special handling for that case. If you push state and click from there to a new page, then reload _that_ page, clicking back twice won't work as expected. Still looking into how to fix this. possibly nuke navigation index history on page reload? unless there's a better way~~ actually i'm not sure what this referred to
- [x] docs
- [x] tests

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: packages/kit/types/index.d.ts [hunk 1]
```diff
@@ -2002,6 +2006,16 @@ declare module '$app/navigation' {
 	 * `afterNavigate` must be called during a component initialization. It remains active as long as the component is mounted.
 	 * */
 	export const afterNavigate: (callback: (navigation: import('@sveltejs/kit').AfterNavigate) => void) => void;
+	/**
+	 * Programmatically create a new history entry with the given `$page.state`. To use the current URL, you can pass `''` as the first argument.
+	 *
```
Comment (benmccann) [inline packages/kit/types/index.d.ts:2011]:
```suggestion
```

File: packages/kit/src/core/sync/write_non_ambient.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ declare module "svelte/elements" {
 			| null;
 		'data-sveltekit-preload-data'?: true | '' | 'hover' | 'tap' | 'off' | undefined | null;
 		'data-sveltekit-reload'?: true | '' | 'off' | undefined | null;
-		'data-sveltekit-replacestate'?: true | '' | 'off' | undefined | null;
+		'data-sveltekit-replace_state'?: true | '' | 'off' | undefined | null;
```
Comment (ignatiusmb) [inline packages/kit/src/core/sync/write_non_ambient.js:29]: Not sure if this change is intended, but should it be a hypen?

```suggestion
		'data-sveltekit-replace-state'?: true | '' | 'off' | undefined | null;
```

Comment (Rich-Harris) [reply packages/kit/src/core/sync/write_non_ambient.js]: oops! no, this was a copy-paste fail

File: packages/kit/src/runtime/app/navigation.js [hunk 3]
```diff
@@ -64,11 +57,11 @@ export const invalidateAll = /* @__PURE__ */ client_method('invalidate_all');
...
  * This is the same behaviour that SvelteKit triggers when the user taps or mouses over an `<a>` element with `data-sveltekit-preload-data`.
  * If the next navigation is to `href`, the values returned from load will be used, making navigation instantaneous.
- * Returns a Promise that resolves when the preload is complete.
+ * Returns a Promise that resolves with the result of running the new route's `load` functions once the preload is complete.
  *
- * @type {(href: string) => Promise<void>}
+ * @type {(href: string) => Promise<Record<string, any>>}
  * @param {string} href Page to preload
- * @returns {Promise<void>}
+ * @returns {Promise<{ type: 'loaded', status: number, data: Record<string, any> } | { type: 'redirect', location: string }>}
```
Comment (ignatiusmb) [inline packages/kit/src/runtime/app/navigation.js:64]: To keep it consistent
```suggestion
 * @returns {Promise<{ type: 'loaded'; status: number; data: Record<string, any> } | { type: 'redirect'; location: string }>}
```

File: packages/kit/src/runtime/client/client.js [hunk 4]
```diff
@@ -28,33 +28,99 @@ import {
...
+import { compact } from '../../utils/array.js';
 import { HttpError, Redirect, SvelteKitError } from '../control.js';
 import { INVALIDATED_PARAM, TRAILING_SLASH_PARAM, validate_depends } from '../shared.js';
-import { INDEX_KEY, PRELOAD_PRIORITIES, SCROLL_KEY, SNAPSHOT_KEY } from './constants.js';
 import { stores } from './singletons.js';
 import { get_message, get_status } from '../../utils/error.js';
 
 let errored = false;
 
+/** @typedef {{ x: number, y: number }} ScrollPosition */
```
Comment (ignatiusmb) [inline packages/kit/src/runtime/client/client.js:49]:
```suggestion
/** @typedef {{ x: number; y: number }} ScrollPosition */
```

File: packages/kit/src/runtime/app/navigation.js [hunk 5]
```diff
@@ -126,3 +119,23 @@ export const onNavigate = /* @__PURE__ */ client_method('on_navigate');
  * @returns {void}
  */
 export const afterNavigate = /* @__PURE__ */ client_method('after_navigate');
+
+/**
+ * Programmatically create a new history entry with the given `$page.state`. To use the current URL, you can pass `''` as the first argument. Used for [shallow routing](https://kit.svelte.dev/docs/shallow-routing).
+ *
+ * @type {(url: string | URL, state: App.PageState) => void}
```
Comment (dummdidumm) [inline packages/kit/src/runtime/app/navigation.js:126]:
```suggestion
 * @type {(url: string | URL; state: App.PageState) => void}
```

File: packages/kit/src/runtime/app/navigation.js [hunk 6]
```diff
@@ -126,3 +119,23 @@ export const onNavigate = /* @__PURE__ */ client_method('on_navigate');
...
+ * @param {string | URL} url
+ * @param {App.PageState} state
+ * @returns {void}
+ */
+export const pushState = /* @__PURE__ */ client_method('push_state');
+
+/**
+ * Programmatically replace the current history entry with the given `$page.state`. To use the current URL, you can pass `''` as the first argument. Used for [shallow routing](https://kit.svelte.dev/docs/shallow-routing).
+ *
+ * @type {(url: string | URL, state: App.PageState) => void}
```
Comment (dummdidumm) [inline packages/kit/src/runtime/app/navigation.js:136]:
```suggestion
 * @type {(url: string | URL; state: App.PageState) => void}
```

**Pull request merged**